File: src/dorkbox/util/cab/structure/CabEnumerator.java
/*
 * Copyright 2012 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.util.cab.structure;

import java.util.Enumeration;
import java.util.NoSuchElementException;

import dorkbox.util.cab.CabDecoder;

public final class CabEnumerator implements Enumeration<Object> {
    private int        fileCount = 0;
    private int        folderCount = 0;

    private CabDecoder cabDecoder;

    private boolean    b;
    private int        folderIndex;

    @Override
    public Object nextElement() {
        if (!this.b) {
            if (this.fileCount < this.cabDecoder.header.cFiles) {
                return this.cabDecoder.files[this.fileCount++];
            }
            throw new NoSuchElementException();
        }

        if (this.cabDecoder.files[this.fileCount].iFolder != this.folderIndex) {
            this.folderIndex = this.cabDecoder.files[this.fileCount].iFolder;

            if (this.folderCount < this.cabDecoder.folders.length) {
                return this.cabDecoder.folders[this.folderCount++];
            }
        }

        if (this.fileCount < this.cabDecoder.header.cFiles) {
            return this.cabDecoder.files[this.fileCount++];
        }

        throw new NoSuchElementException();
    }

    public CabEnumerator(CabDecoder decoder, boolean b) {
        this.cabDecoder = decoder;
        this.b = b;
        this.folderIndex = -2;
    }

    @Override
    public boolean hasMoreElements() {
        return this.fileCount < this.cabDecoder.header.cFiles;
    }
}

File: src/dorkbox/cabParser/structure/CabEnumerator.kt
/*
 * Copyright 2023 dorkbox, llc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package dorkbox.cabParser.structure

import dorkbox.cabParser.CabParser
import java.util.*

class CabEnumerator(private val cabParser: CabParser, private val b: Boolean) : Enumeration<Any> {
    private var fileCount = 0
    private var folderCount = 0
    private var folderIndex: Int

    init {
        folderIndex = -2
    }

    override fun nextElement(): Any {
        if (!b) {
            if (fileCount < cabParser.header.cFiles) {
                return cabParser.files[fileCount++]
            }
            throw NoSuchElementException()
        }
        if (cabParser.files[fileCount].iFolder != folderIndex) {
            folderIndex = cabParser.files[fileCount].iFolder
            if (folderCount < cabParser.folders.size) {
                return cabParser.folders[folderCount++]
            }
        }
        if (fileCount < cabParser.header.cFiles) {
            return cabParser.files[fileCount++]
        }
        throw NoSuchElementException()
    }

    override fun hasMoreElements(): Boolean {
        return fileCount < cabParser.header.cFiles
    }
}
